presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1424,6 +1435,56 @@ else if (materializedViewStatus.isPartiallyMaterialized()) {
             return SqlFormatterUtil.getFormattedSql(unionQuery, sqlParser, Optional.empty());
         }
 
+        /**
+         * Returns materialized view status, considering filter conditions from query
+         */
+        private MaterializedViewStatus getMaterializedViewStatus(QualifiedObjectName materializedViewName, Optional<Scope> scope)
+        {
+            TupleDomain<String> baseQueryDomain;
+            // If there is a filter in the QuerySpecification we are currently visiting which has been rewritten by a materialized view,
+            // We can avoid counting partitions which will not be present in the final query when considering if we fail the optimization
+            // Due to too many missing partitions.
+            // TODO: Collect predicates in outer queries and apply the union to this criteria. i.e. in
```

Comment:
Lets create a github issue for this. Good catch

Reply:
Will do- I'd wait until this is merged, just so any details are unchanging for when I write the issue.

Reply:
This is like reinventing predicate pushdown in analyzer